File: SOLUTIONS/LEETCODE/MEDIUM/findDiagonalOrder/findDiagonalOrder.py
def findDiagonalOrder(mat):
  # number of rows in the matrix
  m = len(mat)
  
  # number of columns in the matrix
  # mat[0] means the very first row, len(mat[0]) counts how many columns that row has
  # this gives us the number of columns for the entire matrix
  n = len(mat[0])
  
  # store for our final diagonal traversal
  store = []
  
  # loop over all diagonals
  # the maximum diagonal index we can get is (m-1) + (n-1) = m + n - 2
  for d in range(m + n - 2):
    packer = [] # temporary storage for this diagonal
    
    # row range calculation
    # rowStart = the lowest row index possible on this diagonal
    #   can't be less than 0 (no negative rows)
    #   also must be at least d - (n - 1) because we can't go beyond the last column
    rowStart = max(0, d - (n -1))
    
    # rowEnd = the highest row index possible on this diagonal
    #   can't be more than m - 1 (last row)
    #   also at most "d" since r + c = d
    rowEnd = min(m - 1, d)
    
    # collect all cells where r + c = d
    for r in range(rowStart, rowEnd + 1):
      c = d - r
      packer.append(mat[r][c])
      
    # for even diagonals, reverse the local order for the zigzag pattern
    if(d % 2 == 0):
      packer.reverse()
      
    # add this diagonal's elements into our store
    store.extend(packer)
    
  return store
  

File: SOLUTIONS/LEETCODE/diagonals/findDiagonalOrder/findDiagonalOrder.js
function findDiagonalOrder(mat) {
  // rows (height of the matrix)
  const m = mat.length
  // columns (using the the length of the first row which is equal to that of other rows)
  const n = mat[0].length
  // create a store for the final traversal
  const result = []

  // formula for diagonal matrices is d = r + c
  // smallest d is [0], top left cell
  // largest d calculation; d = (m-1) + (n-1) = m + n - 2 (bottom right cell)
  for (let d = 0; d <= m + n - 2; d++){
    // empty array to collect iterables as the loop runs
    const packer = []

    // derive valid row range for this diagonal
    // from r >= 0 and c = d - r <= n - 1 => r >= d - (n - 1)
    // from r <= m - 1 and c = d - r >= 0 => r <= d
    // combine and clamp to matrix bounds:
    const rowStart = Math.max(0, d - (n - 1))
    const rowEnd = Math.min(m - 1, d)

    // traverse every (r,c) where d = r + c is valid
    for (let r = rowStart; r <= rowEnd; r++){
      const c = d - r;   // because r + c must equal to d
      packer.push(mat[r][c])
    }

    // zigzag requisite:
    // even d => traverse up right => reverse the local order we just collected
    if (d % 2 === 0) packer.reverse
    
    // append this diagonal to the final result
    result.push(...packer)
  }
  // collect all the numbers in diagonally calculated order
  return result
}